### PR TITLE
Adding e-mail auth (e-mail verfication, password reset)

### PR DESCRIPTION
**Pull request by infomiho**
### E-mail auth

Adds e-mail & password auth method:
- e-mail field validation
- support for e-mail verification (`isEmailVerified`) using JWT token
- support for password reset using JWT token
- sending of e-mails for e-mail verification and password reset
- templating of e-mails
- redirect after e-mail is verified
- route to open when doing password reset

### TODO

We still need to figure out the following:
- [x] can we allow using username & password and e-mail & password at the same time?
  - [x] how can we make that work?
- [x] validate that a e-mail sender is defined if the e-mail & password provider is used  

Closes:
- #891
- #691

### CODE REVIEW

File: waspc/data/Generator/templates/server/src/auth/providers/email/signup.ts
Comment (infomiho) [inline waspc/data/Generator/templates/server/src/auth/providers/email/signup.ts:22]: We need to now know about the backend URL as well and the user needs to provide it through the env variables.

File: waspc/examples/todoApp/src/server/auth/email.ts
Comment (infomiho) [inline waspc/examples/todoApp/src/server/auth/email.ts:6]: User defined e-mail templates

File: waspc/examples/todoApp/migrations/20230328100125_add_email_verified/migration.sql
Comment (shayneczyzewski) [inline waspc/examples/todoApp/migrations/20230328100125_add_email_verified/migration.sql:8]: Do we also want to track other things, like number of emails sent, or last email send time, etc. so we don't allow users to spam emails via our apps?

Comment (shayneczyzewski) [reply waspc/examples/todoApp/migrations/20230328100125_add_email_verified/migration.sql:8]: Also, maybe we should put this on a different object? Unsure, but that would avoid users having to add lots of fields to user if we just use an FK.

File: waspc/examples/todoApp/src/server/auth/email.ts
Comment (shayneczyzewski) [reply waspc/examples/todoApp/src/server/auth/email.ts:6]: In the future maybe we can use some sort of templating for writing the text/html strings more ergonomically. Maybe something simple like loading Handlebars/Mustache templats, or maybe even rendering React components to strings (https://legacy.reactjs.org/docs/react-dom-server.html)?

File: waspc/data/Generator/templates/server/src/auth/providers/email/signup.ts
Comment (shayneczyzewski) [inline waspc/data/Generator/templates/server/src/auth/providers/email/signup.ts:19]: I worry a bit about using the real user model here with pending/to be verified accounts. I think there may be flows that are not desireable (beyond the ability to just spam the system and clutter the user table).

File: waspc/data/Generator/templates/server/src/auth/providers/email/resetPassword.ts
Comment (shayneczyzewski) [inline waspc/data/Generator/templates/server/src/auth/providers/email/resetPassword.ts:8]: How long are tokens valid for? Should we make it shorter than our default expiration?

File: waspc/data/Generator/templates/server/src/auth/providers/email/login.ts
Comment (shayneczyzewski) [inline waspc/data/Generator/templates/server/src/auth/providers/email/login.ts:7]: Don't we also need to make sure they are verified?

File: waspc/data/Generator/templates/server/src/auth/utils.ts
Comment (shayneczyzewski) [inline waspc/data/Generator/templates/server/src/auth/utils.ts:65]: Just seeing this update function made me wonder if we provide any methods to update other user fields. We may want to put some restrictions in auth validations on updating the `User.email` since then they could take over any other email after they verified theirs.

**Review by shayneczyzewski — COMMENTED** [review]
This is cool @infomiho, nice work! I did not yet run it or go through the code in a super detailed way, but rather just made some high-level questions/comments so we can discuss some design considerations as quickly as possible. Happy to hop on a call to chat/brainstorm if helpful. I will start testing and reviewing in more detail tomorrow 👍🏻

File: waspc/data/Generator/templates/server/src/auth/providers/email/signup.ts
Comment (infomiho) [reply waspc/data/Generator/templates/server/src/auth/providers/email/signup.ts:22]: The users won't see this URL at all since it will redirect to the client URL right after it processes the token it received.

In the RFC we had two options:
1. Load the client page first 🔵
    1. Client page loads with `?token`
    2. It sends a backend request with the token
    3. User is verified
2. Go directly to the backend 🟠
    1. Backend loads with `?token`
    2. Backend redirects to a success or failure page
    
  The second 🟠 one seemed simpler at the time. But now that I implemented it, I see that we need to introduce a mandatory env variable called `BACKEND_URL` to get it working properly. Maybe the first 🔵 one is better for the user after all?

File: waspc/examples/todoApp/src/server/auth/email.ts
Comment (infomiho) [reply waspc/examples/todoApp/src/server/auth/email.ts:6]: I agree 100%, we should consider templating for e-mails in some future iteration. Then this will benefit the users when using e-mail auth. There is thing interesting library for creating e-mails with React: https://react.email/docs/utilities/render

File: waspc/data/Generator/templates/server/src/auth/providers/email/requestPasswordReset.ts
Comment (infomiho) [reply waspc/data/Generator/templates/server/src/auth/providers/email/requestPasswordReset.ts:28]: Ideally, we would `await` the result here and do something with the info. If the API call fails, then maybe return a `500` here. But if the API call succeeds, but the e-mail delivery fails then it's a different kind of problem. 

The issue is that various e-mail providers return different kind of "success" messages, and with most you don't get a definite answer other than "it's scheduled for sending successfully".

File: waspc/examples/todoApp/migrations/20230328075236_add_email/migration.sql
Comment (infomiho) [reply waspc/examples/todoApp/migrations/20230328075236_add_email/migration.sql:14]: We can do that 👍 and it makes our existing docs valid for e-mail & password auth for the most part. 

When we tackle the "provider data" issue and make it so that every provider has its own piece of the database, then they would split.

File: waspc/data/Generator/templates/server/src/auth/providers/email/signup.ts
Comment (infomiho) [reply waspc/data/Generator/templates/server/src/auth/providers/email/signup.ts:19]: > using the real user model here with pending/to be verified accounts

I can see a scenario where somebody knows my e-mail and create an account with my e-mail.

1. They create an account with "miho@xyz.com"
2. I get an e-mail (spam)
3. I later on want to create my real account
4. The e-mail is taken and i would need to use "reset password" to get the actual password

That sounds inconvenient, true. But it seems like it would work itself out.  

> beyond the ability to just spam the system and clutter the user table

That's also a problem with username & password auth as well.

File: waspc/data/Generator/templates/server/src/auth/providers/email/resetPassword.ts
Comment (infomiho) [reply waspc/data/Generator/templates/server/src/auth/providers/email/resetPassword.ts:9]: I've updated the code to make the tokens valid for 30 minutes.

File: waspc/data/Generator/templates/server/src/auth/providers/email/login.ts
Comment (infomiho) [reply waspc/data/Generator/templates/server/src/auth/providers/email/login.ts:7]: There are two options we considered:
1. Unverified users have access by default and the developer decides within their actions what to do
2. Unverified users should not be allowed to log in at all

The first flow sounds to me like it could be better with providing extra onboarding steps while the user didn't verify their e-mail. I'm not sure which one we want to go with. Maybe we put it as an option in the Wasp file?

File: waspc/data/Generator/templates/server/src/auth/utils.ts
Comment (infomiho) [reply waspc/data/Generator/templates/server/src/auth/utils.ts:65]: Changing the email should imply that the user must verify the e-mail again. That's true!

Do we want to expose a function "updateUserEmail" for the user to use or somehow do it inside of Prisma middleware? 🤷

File: waspc/data/Generator/templates/server/src/auth/providers/config/email.ts
Comment (infomiho) [reply waspc/data/Generator/templates/server/src/auth/providers/config/email.ts:73]: We mount this router when we are going through the provider configs. We use the `createRouter` to get a router which we setup under the `/auth` route. https://github.com/wasp-lang/wasp/blob/6dbeedca60edb55315be68477d67e34ae7b1cb5c/waspc/data/Generator/templates/server/src/auth/providers/index.ts#L24

File: waspc/data/Generator/templates/server/src/auth/providers/email/resetPassword.ts
Comment (infomiho) [reply waspc/data/Generator/templates/server/src/auth/providers/email/resetPassword.ts:9]: TODO: make sure we redirect to a user defined route when the token is invalid or expired

File: waspc/data/Generator/templates/server/src/auth/providers/email/signup.ts
Comment (infomiho) [reply waspc/data/Generator/templates/server/src/auth/providers/email/signup.ts:19]: The thrid option might be: don't consider an e-mail taken until verified.

Comment (infomiho) [reply waspc/data/Generator/templates/server/src/auth/providers/email/signup.ts:19]: TODO: research how others are doing e-mail verification

File: waspc/data/Generator/templates/server/src/auth/utils.ts
Comment (infomiho) [reply waspc/data/Generator/templates/server/src/auth/utils.ts:65]: Additional option: provide an option `allowEmailModification` which is by default `false`, and based on that flag we implement Prisma middleware.

**Comment by infomiho** [comment]
@shayneczyzewski ready for another round
- added a rate limiter
- uses `username` instead of `email` as field
- if an existing user with unverified e-mail is detected, they are deleted when a new signup is attempted (Supabase auth does it similarly)

**Review by shayneczyzewski — COMMENTED** [review]
Hi @infomiho I spent some time playing with it locally yesterday and today, it is looking better with the most recent changes! Nice job!

Here are some stream of conscious notes I took while testing. I will do more on this Monday, while also working on removing the username & pass coupling with oauth. 👍🏻 

## NOTES

1. Clicking "Sign up" didn't provide any info to user about what was happening. It looks like under the covers it was successful though. :thinking:

   ```sh
   Server: Test email (not sent): {
   Server:   from: { email: 'shayne.czyzewski@gmail.com', name: 'ToDO App' },
   Server:   to: 'thelastemail@gmail.com',
   Server:   subject: 'Verify your email',
   Server:   text: 'Click the link below to verify your email: http://localhost:3000/email-verification-?token=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6MywiaWF0IjoxNjgwMjAyOTE2LCJleHAiOjE2ODAyMDQ3MTZ9.YskUq4e123cHplCXK2yWLXqNaPkC8piFF0pHRuKJwc8',
   Server:   html: '\n' +
   Server:     '        <p>Click the link below to verify your email</p>\n' +
   Server:     '        <a href="http://localhost:3000/email-verification-?token=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6MywiaWF0IjoxNjgwMjAyOTE2LCJleHAiOjE2ODAyMDQ3MTZ9.YskUq4e123cHplCXK2yWLXqNaPkC8piFF0pHRuKJwc8">Verify email</a>\n' +
   Server:     '    '
   Server: }
   Server: POST /auth/email/signup 200 137.200 ms - 16
   Server: POST /auth/email/login 401 1.881 ms - -
   Server: GET /auth/me 401 0.297 ms - -
   ```

   We will probably want to work on the redirects and/or provide a response they can use for a toast message.

   Same with request password reset.

2. We may want to avoid leaking the info that:

      > Error: User with this email already exists. - Details: User with this email already exists.

when you try to sign up with an email in the system. Although I'm not sure what we can say, but this reminds me of some security article I read discouraging that since attackers can then figure out what emails are registered with sites.

3. It appeared like the timestamps about email sending were not being updated in either send or non-send mode.

**Comment by shayneczyzewski** [comment]
Here is what I'm working on to remove the oauth dep with `usernameAndPassword` we will need for this @infomiho: https://github.com/wasp-lang/wasp/pull/1094

File: waspc/data/Generator/templates/server/src/middleware/rateLimiter.ts
Comment (infomiho) [reply waspc/data/Generator/templates/server/src/middleware/rateLimiter.ts:6]: I've decided to not introduce the generic limiter, but only for the two routes we have for the e-mail auth. I want to better support rate limiting customisation for the general case. 

Now we only have a rate limiter for signup and password reset and just let through everything else like before.

Comment (infomiho) [reply waspc/data/Generator/templates/server/src/middleware/rateLimiter.ts:6]: Issue for rate limiting improvement https://github.com/wasp-lang/wasp/issues/1098

File: waspc/data/Generator/templates/server/src/core/auth/prismaMiddleware.js
Comment (infomiho) [reply waspc/data/Generator/templates/server/src/core/auth/prismaMiddleware.js:74]: Since our UI will probably say "E-mail" and the fact that this is called `username` will be an implementation detail, I'd rather keep the existing message.

File: waspc/examples/todoApp/migrations/20230328100331_update_default_value/migration.sql
Comment (infomiho) [reply waspc/examples/todoApp/migrations/20230328100331_update_default_value/migration.sql:2]: I'll clean up the migration 👍 

Thanks for the heads up

**Comment by infomiho** [comment]
@shayneczyzewski this is ready for another round, changes:

- Compiler forbids usage of `username & password` and `email` auth at the same time
- Model for `email` auth uses now **optional** `email` and `password` fields
- Validation of `email` auth fields happens in the routes vs. at the Prisma middleware
- I've tested the following scenarios locally:
    - Signup with Google then login
    - Signup with email then login (verification required & not)
    - Reset password
    - Signup with Google then reset password
    - Signup with email then login with Google ⚠️ (this doesn't work since we aren't doing anything to link the accounts)
 - Adds timestamp whenever we send the verification and password reset e-mails to limit that to **once per minute**
 - General fixes and error handling improvements
 - I've deleted the migrations in the example app and I propose we keep only **one migration file** to make it easier for development 🤷‍♂️

File: waspc/data/Generator/templates/server/src/core/auth/prismaMiddleware.js
Comment (shayneczyzewski) [inline waspc/data/Generator/templates/server/src/core/auth/prismaMiddleware.js:7]: We could have an error if username is on the User entity, but usernameandPassword auth is _not_ enabled (due to use below). Maybe we just don't wrap this?

File: waspc/examples/todoApp/src/server/auth/google.js
Comment (shayneczyzewski) [inline waspc/examples/todoApp/src/server/auth/google.js:6]: Nice! We will probably want to note this in the docs somewhere as well 👍🏻

File: waspc/src/Wasp/Analyzer/StdTypeDefinitions.hs
Comment (shayneczyzewski) [inline waspc/src/Wasp/Analyzer/StdTypeDefinitions.hs:34]: Was there a reason to swap this ordering?

File: waspc/data/Generator/templates/server/src/auth/providers/local/login.ts
Comment (shayneczyzewski) [inline waspc/data/Generator/templates/server/src/auth/providers/local/login.ts:10]: This appears to break `usernameAndPassword` auth.

**Review by shayneczyzewski — COMMENTED** [review]
This is looking great @infomiho! I only found one main bug (just `usernameAndPassword` doesn't work due to an email for local in get user). Otherwise, the testing is coming back clean (including just auth working still)! I will do another round of testing later today 👍🏻 Great work 🥳

File: waspc/data/Generator/templates/react-app/src/auth/email/actions/passwordReset.ts
Comment (infomiho) [reply waspc/data/Generator/templates/react-app/src/auth/email/actions/passwordReset.ts:6]: A great _catch_, I'll remove it!

File: waspc/data/Generator/templates/server/src/auth/providers/email/signup.ts
Comment (infomiho) [reply waspc/data/Generator/templates/server/src/auth/providers/email/signup.ts:31]: I've lowercased everything 👍

File: waspc/data/Generator/templates/server/src/auth/providers/local/login.ts
Comment (infomiho) [reply waspc/data/Generator/templates/server/src/auth/providers/local/login.ts]: Should be fixed now 👍

File: waspc/data/Generator/templates/server/src/core/auth/prismaMiddleware.js
Comment (infomiho) [reply waspc/data/Generator/templates/server/src/core/auth/prismaMiddleware.js:7]: Updated to use `isUsernameOnUserEntity` to keep it consistent.

File: waspc/src/Wasp/Analyzer/StdTypeDefinitions.hs
Comment (infomiho) [reply waspc/src/Wasp/Analyzer/StdTypeDefinitions.hs:34]: Since it's template Haskell it expected `Route` to be defined before `App` because `App` now uses `Ref Route` for the email auth definition, something @Martinsos helped me with.

File: waspc/data/Generator/templates/react-app/src/auth/email/actions/passwordReset.ts
Comment (shayneczyzewski) [inline waspc/data/Generator/templates/react-app/src/auth/email/actions/passwordReset.ts:16]: Just spot checking changed files. Does this need a `.data` too?

**Comment by shayneczyzewski** [comment]
Great job @infomiho all the major stuff and testing looks good! I even tried out a few sample apps and they seem to still work just fine. I think the main discussion points are:
- Should we wait to merge this until the Matija Auth UI stuff lands on `main` and can be integrated into this branch? It is probably ok to combine it all in a separate PR like you mentioned, but `main` will be "dirty" until then and we will need to be pretty certain that can get done quickly to make sure `main` is RC-ready again soon.
- I think we can merge before the docs are ready (as long as we get them done by next week).

🤔 any other considerations before this could be merged? Want to run this by anyone else for 👀 too or just use them for final testing?

Nice work on this! 🥳 Functionally and code wise it feels good to go for me 🚀

File: waspc/data/Generator/templates/server/src/auth/providers/email/signup.ts
Comment (shayneczyzewski) [reply waspc/data/Generator/templates/server/src/auth/providers/email/signup.ts:31]: Nice! The` one place we may want to not do so is `data/Generator/templates/server/src/auth/providers/local/login.ts` if the old way did not, just for backward compatibility.

File: waspc/data/Generator/templates/react-app/src/auth/email/actions/passwordReset.ts
Comment (infomiho) [reply waspc/data/Generator/templates/react-app/src/auth/email/actions/passwordReset.ts:16]: Right you are 👍

File: waspc/data/Generator/templates/server/src/auth/providers/email/signup.ts
Comment (infomiho) [reply waspc/data/Generator/templates/server/src/auth/providers/email/signup.ts:31]: I think this is small change and we are pre 1.0, so if you agree, I'll keep the change

**Review by shayneczyzewski — APPROVED** [review]
approving now, group agrees as do I we can just merge this in and add docs/auth UI after, followed by final testing 👍🏻 nice work! 🥳

**Pull request merged**
